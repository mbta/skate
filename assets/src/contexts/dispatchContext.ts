import { createContext } from "react"
import { Dispatch } from "../state"

// tslint:disable-next-line: no-empty
const noopDispatch: Dispatch = () => {}
const DispatchContext = createContext(noopDispatch)

export default DispatchContext
