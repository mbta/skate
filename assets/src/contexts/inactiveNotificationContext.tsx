import React, { createContext, ReactElement } from "react"
import { Notification } from "../realtime.d"

export const InactiveNotificationContext = createContext<
  [Notification | null, (notification: Notification | null) => void]
>(
  // tslint:disable-next-line no-empty
  [null, () => {}]
)

export const InactiveNotificationProvider = ({
  notification,
  setInactiveNotification,
  children,
}: {
  notification: Notification | null
  setInactiveNotification: (notification: Notification | null) => void
  children: ReactElement<HTMLElement>
}) => {
  return (
    <InactiveNotificationContext.Provider
      value={[notification, setInactiveNotification]}
    >
      {children}
    </InactiveNotificationContext.Provider>
  )
}
