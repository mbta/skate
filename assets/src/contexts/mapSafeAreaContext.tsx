import { createContext } from "react"
import { LeafletPaddingOptions } from "../components/stopCard"

export const MapSafeAreaContext = createContext<LeafletPaddingOptions>({})
