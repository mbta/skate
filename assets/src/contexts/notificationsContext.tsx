import React, { createContext, ReactElement, useState } from "react"
import { useNotifications } from "../hooks/useNotifications"
import { Notification, NotificationId } from "../realtime.d"

export interface NotificationsStatus {
  notifications: Notification[]
  removeNotification: (notificationId: NotificationId) => void
}

export const NotificationsContext = createContext<NotificationsStatus>({
  notifications: [],
  // tslint:disable-next-line: no-empty
  removeNotification: () => {},
})

const deliveryFullstoryEvent = (numStacked: number): void => {
  if (window.FS && window.username) {
    window.FS.event("Notification delivered", {
      num_stacked_int: numStacked,
    })
  }
}

export const NotificationsProvider = ({
  children,
}: {
  children: ReactElement<HTMLElement>
}) => {
  const [notifications, setNotifications] = useState<Notification[]>([])
  const addNotification = (notification: Notification): void => {
    setNotifications((previous) => {
      const newNotifications = [...previous, notification]
      deliveryFullstoryEvent(newNotifications.length)
      return newNotifications
    })
  }
  const removeNotification = (id: NotificationId): void => {
    setNotifications((previous) => previous.filter((n) => n.id !== id))
  }

  useNotifications(addNotification)

  return (
    <NotificationsContext.Provider
      value={{
        notifications,
        removeNotification,
      }}
    >
      {children}
    </NotificationsContext.Provider>
  )
}
