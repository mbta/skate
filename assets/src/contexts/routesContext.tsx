import React, { createContext, ReactElement, useContext } from "react"
import { Route, RouteId } from "../schedule"

const RoutesContext = createContext<Route[] | null>(null)

export const useRoute = (routeId: RouteId | null | undefined): Route | null => {
  const routes: Route[] | null = useContext(RoutesContext)
  return (routes || []).find((route) => route.id === routeId) || null
}

export const RoutesProvider = ({
  routes,
  children,
}: {
  routes: Route[] | null
  children: ReactElement<HTMLElement>
}) => {
  return (
    <RoutesContext.Provider value={routes}>{children}</RoutesContext.Provider>
  )
}

export default RoutesContext
