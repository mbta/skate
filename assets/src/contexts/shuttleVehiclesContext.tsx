import React, { Context, createContext, ReactElement } from "react"
import { Vehicle } from "../realtime"

interface Props {
  shuttles: Vehicle[] | null
  children: ReactElement<HTMLElement>
}

export const ShuttleVehiclesContext: Context<Vehicle[] | null> = createContext<
  Vehicle[] | null
>(null)

export const ShuttleVehiclesProvider = ({ shuttles, children }: Props) => (
  <ShuttleVehiclesContext.Provider value={shuttles}>
    {children}
  </ShuttleVehiclesContext.Provider>
)
