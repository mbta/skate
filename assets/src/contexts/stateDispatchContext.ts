import { createContext } from "react"
import { Dispatch, initialState, State } from "../state"

type StateDispatchContextData = [State, Dispatch]

// tslint:disable-next-line: no-empty
const noopDispatch: Dispatch = () => {}
const StateDispatchContext = createContext([
  initialState,
  noopDispatch,
] as StateDispatchContextData)

export default StateDispatchContext
