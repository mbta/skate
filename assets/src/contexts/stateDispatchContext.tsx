import React, { createContext, ReactElement } from "react"
import { Dispatch, initialState, State } from "../state"

export type StateDispatchContextData = [State, Dispatch]

// tslint:disable-next-line: no-empty
const noopDispatch: Dispatch = () => {}
export const StateDispatchContext = createContext([
  initialState,
  noopDispatch,
] as StateDispatchContextData)

export const StateDispatchProvider = ({
  state,
  dispatch,
  children,
}: {
  state: State
  dispatch: Dispatch
  children: ReactElement<HTMLElement>
}) => {
  return (
    <StateDispatchContext.Provider value={[state, dispatch]}>
      {children}
    </StateDispatchContext.Provider>
  )
}
