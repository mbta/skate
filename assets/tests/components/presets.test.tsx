import React from "react"
import { render } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import renderer from "react-test-renderer"
import {
  initialState,
  createPreset,
  instantiatePreset,
  savePreset,
} from "../../src/state"
import Presets from "../../src/components/presets"
import { StateDispatchProvider } from "../../src/contexts/stateDispatchContext"
import routeTabFactory from "../factories/routeTab"

describe("Presets", () => {
  test("renders current presets", () => {
    const mockDispatch = jest.fn()
    const mockState = {
      ...initialState,
      routeTabs: [
        routeTabFactory.build({
          ordering: 0,
          presetName: "My Preset",
          isCurrentTab: true,
          selectedRouteIds: ["1"],
        }),
        routeTabFactory.build({
          ordering: 1,
          presetName: "Another Preset",
          isCurrentTab: false,
          selectedRouteIds: ["77"],
        }),
        routeTabFactory.build({
          ordering: undefined,
          isCurrentTab: false,
          selectedRouteIds: ["28"],
        }),
      ],
    }
    const tree = renderer
      .create(
        <StateDispatchProvider state={mockState} dispatch={mockDispatch}>
          <Presets />
        </StateDispatchProvider>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("saves current tab as preset", () => {
    jest.spyOn(Math, "random").mockImplementationOnce(() => {
      return 0.01
    })
    const mockDispatch = jest.fn()
    const mockState = {
      ...initialState,
      routeTabs: [
        routeTabFactory.build({
          uuid: "uuid1",
          ordering: 0,
          presetName: undefined,
          isCurrentTab: true,
          selectedRouteIds: ["1"],
        }),
      ],
    }

    const result = render(
      <StateDispatchProvider state={mockState} dispatch={mockDispatch}>
        <Presets />
      </StateDispatchProvider>
    )

    userEvent.click(result.getByText("Save as preset"))

    expect(mockDispatch).toHaveBeenCalledWith(
      createPreset("uuid1", "Preset 100")
    )
  })

  test("saves changes to current tab if it's edited", () => {
    const mockDispatch = jest.fn()
    const mockState = {
      ...initialState,
      routeTabs: [
        routeTabFactory.build({
          uuid: "uuid1",
          ordering: undefined,
          presetName: "My Preset",
          isCurrentTab: false,
          selectedRouteIds: ["1"],
        }),
        routeTabFactory.build({
          uuid: "uuid2",
          ordering: 0,
          presetName: "My Preset",
          isCurrentTab: true,
          selectedRouteIds: ["1", "7"],
          saveChangesToTabUuid: "uuid1",
        }),
      ],
    }

    const result = render(
      <StateDispatchProvider state={mockState} dispatch={mockDispatch}>
        <Presets />
      </StateDispatchProvider>
    )

    userEvent.click(result.getByText("Save as preset"))

    expect(mockDispatch).toHaveBeenCalledWith(savePreset("uuid2"))
  })

  test("opens a preset", () => {
    const mockDispatch = jest.fn()
    const mockState = {
      ...initialState,
      routeTabs: [
        routeTabFactory.build({
          uuid: "uuid1",
          ordering: 0,
          presetName: "My Preset",
          isCurrentTab: true,
          selectedRouteIds: ["1"],
        }),
      ],
    }

    const result = render(
      <StateDispatchProvider state={mockState} dispatch={mockDispatch}>
        <Presets />
      </StateDispatchProvider>
    )

    userEvent.click(result.getByText("My Preset"))

    expect(mockDispatch).toHaveBeenCalledWith(instantiatePreset("uuid1"))
  })
})
