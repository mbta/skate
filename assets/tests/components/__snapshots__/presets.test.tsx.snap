// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Presets renders current presets 1`] = `
<div
  className="m-presets-panel"
>
  <ul>
    <li>
      <div
        className="m-presets-panel__preset-button-container"
      >
        <button
          onClick={[Function]}
        >
          Another Preset
        </button>
      </div>
      <button
        className="m-close-button"
        data-testid="close-button"
        onClick={[Function]}
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </button>
    </li>
    <li>
      <div
        className="m-presets-panel__preset-button-container"
      >
        <button
          onClick={[Function]}
        >
          My Preset
        </button>
      </div>
      <button
        className="m-close-button"
        data-testid="close-button"
        onClick={[Function]}
      >
        <span
          className=""
          dangerouslySetInnerHTML={
            Object {
              "__html": "SVG",
            }
          }
        />
      </button>
    </li>
  </ul>
  <button
    className="m-presets-panel__save-as-preset-button"
    onClick={[Function]}
  >
    <span
      className="m-presets-panel__save-as-preset-button-icon"
      dangerouslySetInnerHTML={
        Object {
          "__html": "SVG",
        }
      }
    />
    Save as preset
  </button>
</div>
`;
